daiquiri/oai/adapter.py

```diff
@@ -48,3 +105,261 @@ def get_sample_identifier(self):
+            return None
+
+    def get_schema_record(self, schema):
+        identifier = self.get_identifier('schemas/%i' % schema.pk)
+        datestamp = schema.updated or schema.published
+        set_spec = 'schema'
+        public = (schema.metadata_access_level == ACCESS_LEVEL_PUBLIC) \
+            and (schema.published is not None)
+
+        return schema.pk, identifier, datestamp, set_spec, public
+
+    def get_table_record(self, table):
+        identifier = self.get_identifier('tables/%i' % table.pk)
+        datestamp = table.updated or table.published
+        set_spec = 'table'
+        public = (table.metadata_access_level == ACCESS_LEVEL_PUBLIC) \
+            and (table.published is not None) \
+            and (table.schema.metadata_access_level == ACCESS_LEVEL_PUBLIC) \
+            and (table.schema.published is not None)
+
+        return table.pk, identifier, datestamp, set_spec, public
+
+
+class DoiMetadataOaiAdapterMixin(MetadataOaiAdapterMixin):
+
+    def get_schema_record(self, schema):
+        identifier = self.get_identifier(schema.doi)
+        datestamp = schema.updated or schema.published
+        set_spec = 'schema'
+        public = (schema.metadata_access_level == ACCESS_LEVEL_PUBLIC) \
+            and bool(schema.published) \
+            and bool(schema.doi)
+
+        return schema.pk, identifier, datestamp, set_spec, public
+
+    def get_table_record(self, table):
+        identifier = self.get_identifier(table.doi)
+        datestamp = table.updated or table.published
+        set_spec = 'table'
+        public = (table.metadata_access_level == ACCESS_LEVEL_PUBLIC) \
+            and bool(table.published) \
+            and bool(table.doi) \
+            and (table.schema.metadata_access_level == ACCESS_LEVEL_PUBLIC) \
+            and bool(table.schema.published)
+
+        return table.pk, identifier, datestamp, set_spec, public
+
+
+class DatalinkOAIAdapter(object):
+
+    tables = settings.DATALINK_TABLES
+
+    datalink_metadata_prefixes = ('oai_dc', 'oai_datacite', 'datacite')
+
+    def get_oai_dc_datalink_serializer_class(self):
+        return import_class('daiquiri.datalink.serializers.DublincoreSerializer')
+
+    def get_oai_datacite_datalink_serializer_class(self):
+        return import_class('daiquiri.datalink.serializers.DataciteSerializer')
+
+    def get_datacite_datalink_serializer_class(self):
+        return import_class('daiquiri.datalink.serializers.DataciteSerializer')
+
+    def get_datalink_list(self):
+        for table in self.tables:
+            schema_name, table_name = table.split('.')
+            rows = DatabaseAdapter().fetch_rows(schema_name, table_name, column_names=['ID', 'access_url'],
+                                                page_size=0, filters={'semantics': '#doi'})
+
+            for ID, access_url in rows:
+                doi = access_url.replace('https://doi.org/', '')
+                yield 'datalink', {'id': int(ID), 'doi': get_doi(access_url)}
```

Comment:
This is an issue since ID is not necessary an integer. Actually it will be almost never one.

Reply:
I see the issue is that `Record.resource_id` needs to be a BigInt.

Reply:
So I guess we need to define a function `get_resource_id_from_datalink_ID` or something similar.

Because in case of a DOI of a single object like a target of enveloppe... we can not impose the ID in the user defined datalink table to be an INT. The standards says it can be any string.

Reply:
We could use the pk instead this is an INT?

Reply:
I changed it to be a char field. I think this is the easiest solution.